File: backend/monitoring/performance.py
"""
Performance Monitoring

Tracks API latency, database query performance, and system metrics.
"""

import time
from typing import Dict, Any, Optional, Callable
from functools import wraps
from datetime import datetime
from contextlib import contextmanager

from backend.logging_config import get_logger
from backend.cache import get, set

logger = get_logger(__name__)


class PerformanceMonitor:
    """Performance monitoring utility."""
    
    def __init__(self):
        self.metrics: Dict[str, list] = {}
    
    def record_latency(
        self,
        endpoint: str,
        method: str,
        latency_ms: float,
        status_code: int = 200,
    ):
        """
        Record API endpoint latency.
        
        Args:
            endpoint: API endpoint path
            method: HTTP method
            latency_ms: Latency in milliseconds
            status_code: HTTP status code
        """
        key = f"{method}:{endpoint}"
        
        if key not in self.metrics:
            self.metrics[key] = []
        
        self.metrics[key].append({
            'latency_ms': latency_ms,
            'status_code': status_code,
            'timestamp': datetime.utcnow().isoformat(),
        })
        
        # Keep only last 1000 measurements
        if len(self.metrics[key]) > 1000:
            self.metrics[key] = self.metrics[key][-1000:]
        
        # Log slow requests
        if latency_ms > 500:
            logger.warning(
                f"Slow request: {key} took {latency_ms:.2f}ms (status: {status_code})"
            )
    
    def get_stats(self, endpoint: str, method: str) -> Dict[str, Any]:
        """
        Get performance statistics for an endpoint.
        
        Args:
            endpoint: API endpoint path
            method: HTTP method
        
        Returns:
            Performance statistics
        """
        key = f"{method}:{endpoint}"
        measurements = self.metrics.get(key, [])
        
        if not measurements:
            return {
                'count': 0,
                'p50': 0,
                'p95': 0,
                'p99': 0,
                'avg': 0,
                'max': 0,
            }
        
        latencies = [m['latency_ms'] for m in measurements]
        latencies.sort()
        
        count = len(latencies)
        p50 = latencies[int(count * 0.5)] if count > 0 else 0
        p95 = latencies[int(count * 0.95)] if count > 0 else 0
        p99 = latencies[int(count * 0.99)] if count > 0 else 0
        avg = sum(latencies) / count if count > 0 else 0
        max_latency = max(latencies) if latencies else 0
        
        return {
            'count': count,
            'p50': round(p50, 2),
            'p95': round(p95, 2),
            'p99': round(p99, 2),
            'avg': round(avg, 2),
            'max': round(max_latency, 2),
        }
    
    def get_all_stats(self) -> Dict[str, Dict[str, Any]]:
        """Get performance statistics for all endpoints."""
        stats = {}
        for key in self.metrics.keys():
            method, endpoint = key.split(':', 1)
            stats[key] = self.get_stats(endpoint, method)
        return stats


# Global performance monitor instance
_performance_monitor = PerformanceMonitor()


def get_performance_monitor() -> PerformanceMonitor:
    """Get global performance monitor instance."""
    return _performance_monitor


def measure_latency(endpoint: str, method: str = "GET"):
    """
    Decorator to measure endpoint latency.
    
    Usage:
        @measure_latency("/api/events", "POST")
        async def create_event(...):
            ...
    """
    def decorator(func: Callable):
        @wraps(func)
        async def async_wrapper(*args, **kwargs):
            start_time = time.time()
            try:
                result = await func(*args, **kwargs)
                status_code = getattr(result, 'status_code', 200) if hasattr(result, 'status_code') else 200
                latency_ms = (time.time() - start_time) * 1000
                _performance_monitor.record_latency(endpoint, method, latency_ms, status_code)
                return result
            except Exception as e:
                latency_ms = (time.time() - start_time) * 1000
                status_code = 500
                _performance_monitor.record_latency(endpoint, method, latency_ms, status_code)
                raise
        
        @wraps(func)
        def sync_wrapper(*args, **kwargs):
            start_time = time.time()
            try:
                result = func(*args, **kwargs)
                status_code = getattr(result, 'status_code', 200) if hasattr(result, 'status_code') else 200
                latency_ms = (time.time() - start_time) * 1000
                _performance_monitor.record_latency(endpoint, method, latency_ms, status_code)
                return result
            except Exception as e:
                latency_ms = (time.time() - start_time) * 1000
                status_code = 500
                _performance_monitor.record_latency(endpoint, method, latency_ms, status_code)
                raise
        
        # Return appropriate wrapper based on function type
        import inspect
        if inspect.iscoroutinefunction(func):
            return async_wrapper
        else:
            return sync_wrapper
    
    return decorator


@contextmanager
def measure_query(query_name: str):
    """
    Context manager to measure database query performance.
    
    Usage:
        with measure_query("get_user_events"):
            events = db.query(Event).filter(...).all()
    """
    start_time = time.time()
    try:
        yield
    finally:
        latency_ms = (time.time() - start_time) * 1000
        if latency_ms > 100:  # Log slow queries
            logger.warning(f"Slow query: {query_name} took {latency_ms:.2f}ms")
        else:
            logger.debug(f"Query: {query_name} took {latency_ms:.2f}ms")

File: frontend/lib/monitoring/performance.ts
/**
 * Frontend Performance Monitoring
 * 
 * Tracks frontend performance metrics: page load time, API latency, etc.
 */

import { logger } from '../logger';

interface PerformanceMetric {
  name: string;
  value: number;
  timestamp: Date;
  metadata?: Record<string, unknown>;
}

class PerformanceMonitor {
  private metrics: PerformanceMetric[] = [];
  private maxMetrics = 1000;

  /**
   * Record a performance metric
   */
  record(name: string, value: number, metadata?: Record<string, unknown>): void {
    const metric: PerformanceMetric = {
      name,
      value,
      timestamp: new Date(),
      metadata,
    };

    this.metrics.push(metric);

    // Keep only last N metrics
    if (this.metrics.length > this.maxMetrics) {
      this.metrics = this.metrics.slice(-this.maxMetrics);
    }

    // Log slow operations
    if (value > 1000) {
      logger.warn(`Slow operation: ${name} took ${value}ms`, metadata);
    } else {
      logger.debug(`Performance: ${name} took ${value}ms`, metadata);
    }
  }

  /**
   * Measure async operation
   */
  async measure<T>(
    name: string,
    fn: () => Promise<T>,
    metadata?: Record<string, unknown>
  ): Promise<T> {
    const start = performance.now();
    try {
      const result = await fn();
      const duration = performance.now() - start;
      this.record(name, duration, metadata);
      return result;
    } catch (error) {
      const duration = performance.now() - start;
      this.record(name, duration, { ...metadata, error: true });
      throw error;
    }
  }

  /**
   * Measure sync operation
   */
  measureSync<T>(
    name: string,
    fn: () => T,
    metadata?: Record<string, unknown>
  ): T {
    const start = performance.now();
    try {
      const result = fn();
      const duration = performance.now() - start;
      this.record(name, duration, metadata);
      return result;
    } catch (error) {
      const duration = performance.now() - start;
      this.record(name, duration, { ...metadata, error: true });
      throw error;
    }
  }

  /**
   * Get performance statistics
   */
  getStats(name: string): {
    count: number;
    avg: number;
    min: number;
    max: number;
    p50: number;
    p95: number;
    p99: number;
  } | null {
    const relevantMetrics = this.metrics.filter((m) => m.name === name);
    if (relevantMetrics.length === 0) {
      return null;
    }

    const values = relevantMetrics.map((m) => m.value).sort((a, b) => a - b);
    const count = values.length;
    const sum = values.reduce((a, b) => a + b, 0);
    const avg = sum / count;
    const min = values[0];
    const max = values[count - 1];
    const p50 = values[Math.floor(count * 0.5)];
    const p95 = values[Math.floor(count * 0.95)];
    const p99 = values[Math.floor(count * 0.99)];

    return {
      count,
      avg: Math.round(avg * 100) / 100,
      min: Math.round(min * 100) / 100,
      max: Math.round(max * 100) / 100,
      p50: Math.round(p50 * 100) / 100,
      p95: Math.round(p95 * 100) / 100,
      p99: Math.round(p99 * 100) / 100,
    };
  }

  /**
   * Get all metrics
   */
  getAllMetrics(): PerformanceMetric[] {
    return [...this.metrics];
  }

  /**
   * Clear metrics
   */
  clear(): void {
    this.metrics = [];
  }
}

// Global instance
export const performanceMonitor = new PerformanceMonitor();

// Web Vitals integration
if (typeof window !== 'undefined') {
  import('web-vitals').then(({ onCLS, onFID, onFCP, onLCP, onTTFB, onINP }) => {
    onCLS((metric) => {
      performanceMonitor.record('web-vitals-cls', metric.value, {
        rating: metric.rating,
      });
    });

    onFID((metric) => {
      performanceMonitor.record('web-vitals-fid', metric.value, {
        rating: metric.rating,
      });
    });

    onFCP((metric) => {
      performanceMonitor.record('web-vitals-fcp', metric.value, {
        rating: metric.rating,
      });
    });

    onLCP((metric) => {
      performanceMonitor.record('web-vitals-lcp', metric.value, {
        rating: metric.rating,
      });
    });

    onTTFB((metric) => {
      performanceMonitor.record('web-vitals-ttfb', metric.value, {
        rating: metric.rating,
      });
    });

    onINP((metric) => {
      performanceMonitor.record('web-vitals-inp', metric.value, {
        rating: metric.rating,
      });
    });
  });
}
